rllib/algorithms/bc/bc_rl_module.py

```diff
@@ -0,0 +1,53 @@
+import abc
+from typing import Any, Type
+
+from ray.rllib.core.models.specs.typing import SpecType
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.models.distributions import Distribution
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import ExperimentalAPI, override
+
+
+@ExperimentalAPI
+class BCRLModule(RLModule, abc.ABC):
+    @override(RLModule)
+    def setup(self):
+        # __sphinx_doc_begin__
+        catalog = self.config.get_catalog()
+
+        # Build models from catalog
+        self.encoder = catalog.build_encoder(framework=self.framework)
+        self.pi = catalog.build_pi_head(framework=self.framework)
+
+        self.action_dist_cls = catalog.get_action_dist_cls(framework=self.framework)
+
+    @override(RLModule)
+    def get_train_action_dist_cls(self) -> Type[Distribution]:
+        return self.action_dist_cls
+
+    @override(RLModule)
+    def get_exploration_action_dist_cls(self) -> Type[Distribution]:
+        return self.action_dist_cls
+
+    @override(RLModule)
+    def get_inference_action_dist_cls(self) -> Type[Distribution]:
+        return self.action_dist_cls
+
+    @override(RLModule)
+    def get_initial_state(self) -> Any:
+        if hasattr(self.encoder, "get_initial_state"):
+            return self.encoder.get_initial_state()
+        else:
+            return {}
+
+    @override(RLModule)
+    def output_specs_inference(self) -> SpecType:
```

Comment:
@kouroshHakha For the specs of all of these RL Modules, I think they should just return a `Spec`, right?

Right now, we have a mixture of `SpecType` and `SpecDict` which are both not correct.
This should be any Spec, just something that I can call `Spec.validate` on.

Reply:
Following the example [here](https://docs.ray.io/en/releases-2.6.3/rllib/rllib-rlmodule.html#writing-custom-single-agent-rl-modules): 
```python
class DiscreteBCTorchModule(TorchRLModule):
    ...

    @override(TorchRLModule)
    def input_specs_exploration(self) -> SpecType:
        # Enforce that input nested dict to exploration method has a key "obs"
        # and within that key, it has a key "global" and "local". There should
        # also be a key "action_mask"
        return [("obs", "global"), ("obs", "local"), "action_mask"]
```
and the deifnition of `format_to_canonical_form()`.

The output of `output_spec_inference()` can be of a type that does not inherit from `Spec` as I understand it and therefore the `SpecType` is used that also allows a list of string or tuple of strings. 

We could possible use instead of the "simpler" list of string here, a `SpecDict` as the `SampleBatch.ACTION_DIST_INPUTS` will be `TensorType`s.

Reply:
If I remember correctly SpecType is a superset of all possible format that we support which includes Spec and SpecDicts and normal dicts, etc.

Reply:
@kouroshHakha Thanks for the clarification!

Reply:
Ok so we should go with SpecType in the general case.